### PR TITLE
SP-6318 prometheus refactor

### PR DESCRIPTION
**Pull request by nshem**

### CODE REVIEW

File: steps/prometheus/query/range/range_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+			Image: testImage,
+			Envs: map[string]string{
+				"URL":   prometheusInstance.ServiceURL.FullURL,
+				"QUERY": "go_gc_duration_seconds",
+				"START": "2021-05-26T19:42:26+00:00",
+				"END":   "2021-05-26T20:42:20+00:00",
+			},
+			Args:         []string{},
+			WantExitCode: exec.ExitCodeOK,
+			WantOutput:   matcher.All(matcher.TextContains("go_gc_duration_seconds"), matcher.StepStatus(codes.OK, "", "")),
```
Comment (oriser) [inline steps/prometheus/query/range/range_integration_test.go:52]: Do you have some metrics you can check here?

File: steps/prometheus/query/raw/raw_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		{
+			Name:  "No time - use now",
+			Image: testImage,
+			Envs: map[string]string{
+				"URL":   prometheusInstance.ServiceURL.FullURL,
+				"QUERY": "go_gc_duration_seconds",
+			},
+			Args:         []string{},
+			WantExitCode: exec.ExitCodeOK,
+			WantOutput:   matcher.All(matcher.TextContains("go_gc_duration_seconds"), matcher.StepStatus(codes.OK, "", "")),
```
Comment (oriser) [inline steps/prometheus/query/raw/raw_integration_test.go:38]: Do you have some metrics you can check here?

**Review by oriser — COMMENTED** [review]
Looks good! See comments and:
* Change to go 1.16 both in Docker base image and in go.mod
* Common test function should be under `testdata` folder (go compiler ignores this folder and won't compile it into the final binary, we will just accessible to it during tests"

File: steps/prometheus/query/range/range.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+		return stat
+	}
+
+	return status.OK
+}
+
+func (p *PrometheusRangeQuery) parseTimeRange() status.Status {
+	startTime, err := time.Parse(query.TimeArgsFormat, p.args.StartTime)
+	if err != nil {
+		return status.FromError(codes.InvalidArgument, "parsing start time", err)
```
Comment (hagaishapira) [inline steps/prometheus/query/range/range.go:60]:
```suggestion
		return status.FromError(codes.InvalidArgument, "invalid start time", err)
```

File: steps/prometheus/query/range/range.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	}
+	p.parsedStartTime = startTime
+
+	if p.args.isRelativeTimeArgs() {
+		p.parsedEndTime = startTime.Add(p.args.Since)
+	} else {
+		endTime, err := time.Parse(query.TimeArgsFormat, p.args.EndTime)
+		if err != nil {
+			return status.FromError(codes.InvalidArgument, "parsing end time", err)
```
Comment (hagaishapira) [inline steps/prometheus/query/range/range.go:69]:
```suggestion
			return status.FromError(codes.InvalidArgument, "invalid end time", err)
```

File: steps/prometheus/query/range/range.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+func (p *PrometheusRangeQuery) Run() (step.Output, step.Status) {
+	var rangeOutput model.Value
+	var stat step.Status
+
+	rangeOutput, stat = p.QueryRange()
+	if stat.Failed() {
+		return nil, stat
+	}
+
+	stepOutput := make(step.Output)
```
Comment (hagaishapira) [inline steps/prometheus/query/range/range.go:104]: why do you need to do a make and not have a literal assignment ?

File: steps/prometheus/query/raw/raw.go [hunk 6]
```diff
@@ -1,52 +1,72 @@
...
 	}
 
-	evaluationTime := time.Time{}
-	if args.Time != "" {
-		t, err := time.Parse(time.RFC3339, args.Time)
+	if p.args.Time != "" {
+		t, err := time.Parse(query.TimeArgsFormat, p.args.Time)
 		if err != nil {
-			return fmt.Errorf("evaluation time parse failed: %w", err)
+			return status.FromError(codes.InvalidArgument, "parsing time", err)
```
Comment (hagaishapira) [inline steps/prometheus/query/raw/raw.go:34]:
```suggestion
			return status.FromError(codes.InvalidArgument, "invalid time argument", err)
```

File: steps/prometheus/query/raw/raw.go [hunk 7]
```diff
@@ -1,52 +1,72 @@
...
+	api, stat := base.ClientAPI(p.args.URL)
+	if stat.Failed() {
+		return nil, stat
 	}
 
-	output, err := json.Marshal(res)
+	val, _, err := api.Query(context.Background(), p.args.Query, p.parsedTime)
 	if err != nil {
-		return fmt.Errorf("prometheus output marshaling failed: %w", err)
+		return nil, status.FromError(codes.Internal, "query time", err)
```
Comment (hagaishapira) [inline steps/prometheus/query/raw/raw.go:52]: This error message is not friendly enough. Does api.Query return different kinds of errors? can we map those to more friendly messages?

File: steps/prometheus/query/range/range.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+		return nil, stat
+	}
+
+	val, _, err := api.QueryRange(context.Background(), p.args.Query, promv1.Range{
+		Start: p.parsedStartTime,
+		End:   p.parsedEndTime,
+		Step:  p.args.Step,
+	})
+	if err != nil {
+		return nil, status.FromError(codes.Internal, "query range", err)
```
Comment (hagaishapira) [inline steps/prometheus/query/range/range.go:89]: Here as well, the message here is not friendly to a user. We should map query errors to friendly messages which the user can act upon.

**Review by hagaishapira — CHANGES_REQUESTED** [review]

File: steps/prometheus/query/range/range.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+func (p *PrometheusRangeQuery) Init() step.Status {
+	if parseStatus := env.Parse(&p.args); parseStatus.Failed() {
+		return parseStatus
+	}
+
+	if !p.args.isRelativeTimeArgs() && !p.args.isAbsoluteTimeArgs() {
+		return status.FromError(codes.InvalidArgument, "at least one of 'SINCE' or 'END' args cannot be empty", fmt.Errorf(""))
+	}
```
[the remark this answers is not in the document]
Comment (nshem) [reply steps/prometheus/query/range/range.go:48]: done it cause im using isRelativeTimeArgs in another place + readability

File: steps/prometheus/query/range/manifest.yaml [hunk 10]
```diff
@@ -35,11 +35,8 @@ envs:
...
     type: "string"
     example: "\"3d\", \"1h30m\", \"15m30s\""
-  - name: DEBUG
-    description: "Print debug outputs."
-    type: "bool"
-    example: "\"true\",\"false\""
 outputs:
-  - name: basic-logs 
-    description: "Raw output of Prometheus query."
+  - name: api_object
```
[the remark this answers is not in the document]
Comment (nshem) [reply steps/prometheus/query/range/manifest.yaml:39]: its not. armis are using it but a version that is in a branch (us-docker.pkg.dev/stackpulse/public/prometheus/query/raw:SP-2345)

File: steps/prometheus/query/range/range.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+		return nil, stat
+	}
+
+	val, _, err := api.QueryRange(context.Background(), p.args.Query, promv1.Range{
+		Start: p.parsedStartTime,
+		End:   p.parsedEndTime,
+		Step:  p.args.Step,
+	})
+	if err != nil {
+		return nil, status.FromError(base.MapApiErrorCode(err), "query range", err)
```
Comment (hagaishapira) [inline steps/prometheus/query/range/range.go:88]: So other than mapping the code which is great we want to map the message and even add the error object in the verbose, just like we do in okta steps for example.
image

File: steps/prometheus/query/raw/raw.go [hunk 12]
```diff
@@ -1,52 +1,73 @@
...
+	api, stat := base.ClientAPI(p.args.URL)
+	if stat.Failed() {
+		return nil, stat
 	}
 
-	output, err := json.Marshal(res)
+	val, _, err := api.Query(context.Background(), p.args.Query, p.parsedTime)
 	if err != nil {
-		return fmt.Errorf("prometheus output marshaling failed: %w", err)
+		return nil, status.FromError(base.MapApiErrorCode(err), "query time", err)
```
Comment (hagaishapira) [inline steps/prometheus/query/raw/raw.go:52]: So other than mapping the code which is great we want to map the message and even add the error object in the verbose, just like we do in okta steps for example.
image

**Review by hagaishapira — COMMENTED** [review]
See just the error behaviour explicitness

File: steps/prometheus/base/prometheus.go [hunk 13]
```diff
@@ -1,36 +1,44 @@
...
+	URL string `env:"URL,required"`
 }
 
-func Client(args *Args) (prometheus.Client, error) {
-	client, err := prometheus.NewClient(prometheus.Config{Address: args.URL})
+func ClientAPI(URL string) (promv1.API, status.Status) {
+	client, err := prometheus.NewClient(prometheus.Config{Address: URL})
 	if err != nil {
-		return nil, fmt.Errorf("failed creatung prometheus client: %w", err)
+		return nil, status.FromError(codes.Unavailable, "failed creatung prometheus client", err)
```
Comment (hagaishapira) [inline steps/prometheus/base/prometheus.go:17]:
```suggestion
		return nil, status.FromError(codes.Unavailable, "failed creating prometheus client", err)
```

**Review by hagaishapira — APPROVED** [review]
Merge after type fix

**Review by oriser — APPROVED** [review]
Nice one small comment

**Pull request merged**